Engineering/Debugging in JavaScript.md

```diff
@@ -0,0 +1,124 @@
+![Chrome DevTools's console tab](https://wd.imgix.net/image/admin/Diu3Bq4TbPWb9Y5gr7HX.png?auto=format)
+
+### `console.log()`
+The `console.log()`  method writes to the web console. The message can be a single text (with optional replacement values) or any number of JavaScript objects.
+
+#### Syntax
+```js
+console.log(obj1)
+console.log(obj1, /* …, */ objN)
+console.log(msg)
+console.log(msg, subst1, /* …, */ substN)
+```
+
+#### Use-case
+This method can be used to print any kind of variables defined before in it or to just print any message that needs to be displayed to the user.
+```js
+const name = "Dwarves Foundation";
+console.log(name);
+//=> Dwarves Foundation
+```
+
+### `console.assert()`
+The `console.assert()` method writes a message to the console if an expression evaluates to `false`.
+
+#### Syntax
+```js
+console.assert(_expression_, _message_);
+```
+
+#### Use-case
+Let's say you wanted to check for a condition of a user id, you might be checking using the `console.log()` like this:
+```js
+if (!user.id){
+	console.log("User does not exist!") \
+	// user.id = false? => User does not exist!
+}
+```
+Instead, you can check the condition using the `console.assert()`:
+```js
+console.assert(user.id, "User does not exist!") 
+// user.id = false? => Assertion failed: User does not exist!
+```
+Using `console.assert()` provide a much cleaner and better way for conditional logging in your application.
+
+### `console.dir()`
+The `console.dir()` displays all of the properties of a specified JavaScript object in the console, allowing the developer to conveniently access the object's properties.
+
+#### Syntax
+```js
+console.dir(object);
+```
+
+#### Use-case
+In Chrome, `console.log` prints out a tree -- _most of the time_. However, Chrome's `console.log` still stringifies certain classes of objects, even if they have properties. A regular expression is the most obvious example of a distinction:
+```js
+const array = [1,2,3];
+console.log(array);
+// [1, 2, 3]
+console.dir(array);
+/* Array[3]
+    0: 1
+    1: 2
+    2: 3
+    length: 3
+    * __proto__: Array[0]
+        concat: function concat() { [native code] }
+        constructor: function Array() { [native code] }
+        entries: function entries() { [native code] }
+        ... */
+```
+Another useful difference in Chrome exists when sending **`DOM`** elements to the console:
+![console.dir() example](https://pbs.twimg.com/media/Di30ycpW0AAYOx7.jpg:medium)
+
+### `console.time()` and `console.timeEnd()`
+The `console.time()` method launches a timer that you may use to track the duration of the operation. You may have up to 10,000 timers running on a single page, giving each one a unique name. When you use `console.timeEnd()` with the same name, the browser returns the time in milliseconds since the timer was started.
+
+#### Syntax
+```js
+time(label)
```

Comment:
does this mean `console.time(label)`?

Reply:
should add `console.timeEnd(label)` too

Reply:
It is just a label to give the timer a name and this is optional.

Reply:
I mean line 88 syntax should be `console.time(label)`, right?